nanoFramework.CoreLibrary/System/SpanByte.cs

```diff
@@ -0,0 +1,172 @@
+﻿//
+// Copyright (c) .NET Foundation and Contributors
+// Portions Copyright (c) Microsoft Corporation.  All rights reserved.
+// See LICENSE file in the project root for full license information.
+//
+
+namespace System
+{
+    /// <summary>
+    /// Provides a type- and memory-safe representation of a contiguous region of arbitrary byte array
+    /// </summary>
+    public struct SpanByte
+    {
+        private byte[] _array;
+        private readonly int _start;
+        private readonly int _length;
+
+        /// <summary>
+        /// Creates a new System.SpanByte object over the entirety of a specified array.
+        /// </summary>
+        /// <param name="array">The array from which to create the System.Span object.</param>
+        public SpanByte(byte[] array)
+        {
+            _array = array;
+            _length = array != null ? array.Length : 0;
+            _start = 0;
+        }
+
+        /// <summary>
+        /// Creates a new System.SpanByte object that includes a specified number of elements
+        /// of an array starting at a specified index.
+        /// </summary>
+        /// <param name="array">The source array.</param>
+        /// <param name="start">The index of the first element to include in the new System.Span</param>
+        /// <param name="length">The number of elements to include in the new System.Span</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// array is null, but start or length is non-zero. -or- start is outside the bounds
+        /// of the array. -or- start and length exceeds the number of elements in the array.
+        /// </exception>
+        public SpanByte(byte[] array, int start, int length)
+        {
+            if (array != null)
+            {
+                if ((length > array.Length - start) || (start >= array.Length))
+                {
+                    throw new ArgumentOutOfRangeException($"Array length too small");
+                }
+            }
+            else
+            {
+                if ((start != 0) || (length != 0))
+                {
+                    throw new ArgumentOutOfRangeException($"Array is null but start and length are not 0");
+                }
+            }
+
+            _array = array;
+            _start = start;
+            _length = length;
+        }
+
+        /// <summary>
+        /// Gets the element at the specified zero-based index.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element.</param>
+        /// <returns>The element at the specified index.</returns>
+        public byte this[int index]
```

Comment:
After fixing nanoframework/Home#641, it's OK to have this as you had it before:

`public ref byte this[int index] => ref _array[index];`

It looks cleaner and elegant. 😉 

(but it's OK to leave it as it is now, of course...)

Reply:
right :-) Will make it elegant again :-D

Reply:
Well, actually I can't! I need to check the index. So found another element to adjust:
```csharp
                if (_start + index > _length)
                {
                    throw new ArgumentOutOfRangeException($"Index out of range");
                }
```

Reply:
No worries there, that's "checked" at the execution engine. If you try to access an element that doesn't exist it will throw the exception.

Reply:
correct but as the array can be longer, idea is to restrict accessing what has been sliced. Now we can as well take the decision not to do it for performance reasons

Reply:
Won't `public ref byte this[int index] => ref _array[_start + index];` will do just that?

Reply:
> correct but as the array can be longer, idea is to restrict accessing what has been sliced. Now we can as well take the decision not to do it for performance reasons

The code in the .NET Core is making a check that requires one operation less and it works in this situation too.
Would you agree?  https://github.com/dotnet/runtime/blob/e0c2ba7e14eceb0e26173d48417f8081b1a4e8f3/src/libraries/System.Private.CoreLib/src/System/Span.cs#L141

On performance: I think this is quite all right.

Reply:
You are actually right. It should be just ```index > _length```. My bad. Not awake today! Clearly not awake. Now if index = -1 and we have a start = 1, we will get a value while we should not. It's not catastrophic anyway but it's a behavior that can happen.